app/controllers/thirdparty/EoriNumberController.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2025 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.thirdparty
+
+import controllers.actions.*
+import forms.thirdparty.EoriNumberFormProvider
+import models.Mode
+import navigation.ThirdPartyNavigator
+import pages.thirdparty.EoriNumberPage
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import repositories.SessionRepository
+import services.TradeReportingExtractsService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.thirdparty.EoriNumberView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class EoriNumberController @Inject() (
+  override val messagesApi: MessagesApi,
+  sessionRepository: SessionRepository,
+  navigator: ThirdPartyNavigator,
+  identify: IdentifierAction,
+  getData: DataRetrievalAction,
+  requireData: DataRequiredAction,
+  formProvider: EoriNumberFormProvider,
+  tradeReportingExtractsService: TradeReportingExtractsService,
+  val controllerComponents: MessagesControllerComponents,
+  view: EoriNumberView
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport {
+
+  val form = formProvider()
+
+  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData) { implicit request =>
+
+    val preparedForm = request.userAnswers.get(EoriNumberPage) match {
+      case None        => form
+      case Some(value) => form.fill(value)
+    }
+
+    Ok(view(preparedForm, mode))
+
+  }
+
+  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode))),
+          eori =>
+            tradeReportingExtractsService.getCompanyInformation(eori).flatMap { companyInfo =>
+              if (companyInfo.consent.isEmpty) {
+                val formWithApiError = form.withError("value", "eoriNumber.error.notFound")
+                Future.successful(BadRequest(view(formWithApiError, mode)))
+              } else {
+                for {
+                  updatedAnswers <- Future.fromTry(request.userAnswers.set(EoriNumberPage, eori))
+                  _              <- sessionRepository.set(updatedAnswers)
```

Comment:
Can we introduce saving the navigation fragments here in the controllers? Please see the onSubmit method of the ThirdPartyDataOwnerConsentController for an example :)